extensions/MasterMath/midi.js

```diff
@@ -0,0 +1,246 @@
+// Name: MIDI
+// ID: midi
+// Description: An extension that retrieves input from MIDI devices.
+// By: -MasterMath- <https://scratch.mit.edu/users/-MasterMath-/>
+// License: MPL-2.0
+
+(function (Scratch) {
+  "use strict";
+
+  if (!Scratch.extensions.unsandboxed) {
+    alert("This MIDI extension must run unsandboxed!");
+    throw new Error("This MIDI extension must run unsandboxed!");
+  }
+
+  let midiInputDevices = [];
+  let midiDeviceInfo = [];
+  let notesOn = [];
+  let noteVelocities = [];
+  let lastNotePressed = 0;
+  let lastNoteReleased = 0;
+
+  if (navigator.requestMIDIAccess) {
+    navigator.requestMIDIAccess().then(onSuccess, onError);
+
+    function onSuccess(midiAccess) {
+      midiAccess.onstatechange = (event) => {
+        if (event.port.state == "connected") {
+          midiInputDevices.push([
+            `[id: "${event.port.id}"` + ` name: "${event.port.name}"]`,
+          ]);
+          midiDeviceInfo.push([event.port.id, event.port.name]);
+        } else if (event.port.state == "disconnected") {
+          midiInputDevices.splice(
```

Comment:
splice[ takes an index](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice), not an object. I'm guessing you mean something like:
```js
// find index in array of item that matches callback
const index = midiDeviceInfo.findIndex(item => item[0] === event.port.id);
// if found then cut out element at that position
if (index !== -1) midiDeviceInfo.splice(index, 1);
```

Reply:
Yes, when the MIDI device is disconnected it's supposed to remove the corresponding object from the array. Thanks for letting me know, I always miss these little oversights.